stdlib/public/core/Arrays.swift.gyb

```diff
@@ -2297,6 +2297,20 @@ extension ${Self} : Equatable where Element : Equatable {
   }
 }
 
+extension ${Self}: Hashable where Element: Hashable {
+  @_inlineable // FIXME(sil-serialize-all)
+  public var hashValue: Int {
+    return _hashValue(for: self)
+  }
+
+  @_inlineable // FIXME(sil-serialize-all)
+  public func _hash(into hasher: inout _Hasher) {
+    for element in self {
+      hasher.append(element)
+    }
```

Comment:
It's a bit surprising to have the hash be O(n) and dependent on the contents.   `NSDictionary` and friends have to take great care about this (in fact our `-hash` impl currently returns just `-count`) because while we do rely on `NSCopying` for our keys -- our values are allowed to mutate.  In fact thats perhaps a very common use case.

Reply:
Thinking further, that might preclude being able to do this conditional conformance for Array and or Dictionary, unless there is a way to constrain mutability.   Value types are fine obviously,  but if you have reference keys, or reference values, you can't guarantee the conformance.

Reply:
Content-based hashing is very much intentional here; we consider this a feature, not a bug. The built-in hashing behavior is designed to produce robust hash values: if two instances aren't equal, they are expected to produce different hashes. (Note that `String` hashing is also dependent on the length of the string; in fact, it also performs Unicode normalization.) 

The mutation issue is indeed important; however, Swift approaches it in a different way, by concentrating on first-class value types. The language itself makes it impossible to mutate value-typed keys in a `Dictionary`, and the same mechanism protects a `Dictionary`'s values, if the dictionary is put in a `Set`.

As a rule, Swift does not provide language/API-level protection against users mutating **reference-typed** components in a `Dictionary` key (or `Set` element) in such a way that changes the key's equality or hashing behavior. (`Dictionary` and `Set` doesn't even call `-copyWithZone:` for keys that happen to implement `NSCopying`.) This PR merely follows the existing design established for hashing collections.

Note that `Dictionary` only conforms to `Hashable` when both its `Key` and `Value` does; `Hashable` conformance is not automatic for reference types, and random model objects will not get unexpectedly hashed after this PR.

Reply:
I'm actually kind of with Michael on this one. Hashing Array shouldn't take linear time.

Reply:
(After all this I'm not sure it's worth it to make the collections Hashable at all. Optional, sure.)

Reply:
I get that value types don't suffer from this, and also that this only comes into play your underlying types are `Hashable`.  General think if you're a mutable reference type, you shouldn't be hashable necessarily (maybe you `has-a` some kind hashing identity or something).

Swift's focus is value types, yes -- but it also has reference types.  This is a lot of rope for folks to encounter the hard way some day with very little support to statically help find the problem.  But you're right thats the established state of things for now. 

Content based hashing is mathematically very nice.  The perf impact seems problematic though.  Consider a deeply nested JSON / plist, content oriented hashing quickly turns into `O(n ^ x)` especially lacking a memoized variant of the hash.  `String`'s Unicode normalization has a similar (if slightly more finite) problem.    This is especially when you consider you get to do it again for the Equatable check on collision.

Reply:
An additional observation: every successful lookup of a key in a hash table necessarily involves a successful equality check, which is O(n) for collection types like `Array`, `Set` and `Dictionary`. Therefore, O(n) hashing does not change our expected asymptotic performance, but it does prevent the formation of long collision chains that would make the hash table itself do lookups in time that's linear in the size of the table itself.

Reply:
I'm probably missing something crucial, but I fail to get the JSON example; JSON only allows string keys. Is it possible to use deeply nested structures as *dictionary keys* in plists? Why would someone do that?

Putting a deeply nested value type with thousands of components into a `Set` may or may not be slow -- if you actually need the value semantics offered by Swift's standard collection types, then I don't really see how we can be faster. Looking up a value in such a set necessarily involves a successful equality check, which will need to traverse the two structures anyway. Having another traversal for the hash value is merely an additional constant factor. 

If we were to replace the hash value with the collection's count, we would make dictionary lookups quadratic -- equality checks will still be O(n), but we'd have to repeat them for each element in a collision chain, whose length can easily be proportional to the size of the hash table.

There's nothing wrong with using `Set`s as `Dictionary` keys, and there's nothing wrong with using `Array`s as elements in a `Set`. Sure, a hash table may not always be the most appropriate data structure if its elements are large collections -- just like an `Array` is not always appropriate for use as a sorted collection.

Incidentally, we did touch upon this issue at the very end of the original swift-evolution thread for this feature:
https://forums.swift.org/t/let-optional-dictionary-and-array-conditionally-conform-to-hashable/9046/58

It may be a good idea to continue this discussion on that thread rather than in review comments -- not many people follow discussions in random PRs! 😉

Reply:
I see your point regarding keys.  The common cases (outside of just a String/Int/UUID) I've seen there for folks is using a small dictionary of string-string pairs, or an array (ordered uuids).  

Using just `-count` is horrible, I agree.  We want to do something better even for the Foundation types. My point is, if there is a way to make `-hash` less expensive but still provide some uniqueness information that's an improvement.

Apologies for missing the SE thread, its so rare I ever get bandwidth to focus on Swift. :( I'll take a look there for context.

Reply:
To be clear, better protection against mutations in reference types and exploring performance implications of content-based hashing are both valid points that deserve discussion. We did not spend enough time on these on swift-evolution, so it's good we are at least discussing them now.

I really don't have a good answer to the mutations issue - this PR fits comfortably within the existing Swift model, but perhaps we should do more. Standalone Swift classes aren't Hashable by default, which alleviates the issue a little, but not by much. One thing Set and Dictionary may do is to at least detect invalid mutations sometime after they happen; for example, while they're rehashing elements. (Although it's unclear if a delayed error/trap would be helpful or just confusing.) Would some language level change help more? This is far beyond the scope of this PR, but it's something to ponder. 

The performance concerns seem easier to tackle, because we can run benchmarks. For the small cases you mentioned, I’m absolutely positive that hashing the full contents is the only reasonable thing we can do - leaving even a mere handful of relevant bits out of the hash would open the door to runaway collision chains. (Beyond security/reliability concerns, this may also have an implication for storage overhead: I have plans to try increasing the maximum load factor from 75% to around 95% at some point, but this requires extremely robust hashing, as well as adding some extra bits to hash table entries.) 

We may get away with sampling a random (but stable) fraction of Array elements when we need to hash large arrays, but given that we can't apply the same trick in ==, it’s not at all clear to me if that would make a difference that's worth the reliability costs. Similarly, Sets could maintain a running hash of their contents — the extra tradeoff there is that it would add overhead to set mutations for all sets, including the vast majority that will never be hashed. (The same applies to Dictionary.) In any case, we'll be able to tweak the implementations in this PR later; hash values are explicitly not stable across even executions of the same Swift program, and we can change the algorithm that generates them in future versions of the stdlib. (As long as we remove the inlinable attribute, which we will probably do as part of an ABI audit.)

Hashing n bits takes O(n) time; this is just how hashing works. Consequently, hashing a list of n elements takes n times slower than a single element. This cost may be acceptable for some applications, and prohibitive to others, but this can be said of every single operation offered by the stdlib. I don't see why we would single out Array as in need of constant-time hashing, while at the same time, we're perfectly fine with String and Data taking linear time.

1. There are clear usecases for hashing at least small arrays/sets/dictionaries, 
2. Content-based hashing is not only the most robust choice, but it also aligns perfectly with Swift's emphasis on collections that are themselves value types, and how these collections already implement Equatable. (After all, hashValue is an approximation for one half of a ==...)
3. Users can easily implement hash memoization and similar techniques to manually speed up hashing when necessary.

Reply:
This discussion has pushed me to considering finally fixing the low numbered radar about collection -hash implementation being so poor, then NS.* and Swift will be consistent together.  Thanks Karoy!